compiler/rustc_middle/src/ty/offload_meta.rs

```diff
@@ -0,0 +1,70 @@
+use crate::ty::{self, PseudoCanonicalInput, Ty, TyCtxt, TypingEnv};
+
+// TODO(Sa4dUs): it doesn't feel correct for me to place this on `rustc_ast::expand`, will look for a proper location
+pub struct OffloadMetadata {
+    pub payload_size: u64,
+    pub mode: TransferKind,
+}
+
+pub enum TransferKind {
+    FromGpu = 1,
+    ToGpu = 2,
+    Both = 3,
+}
+
+impl OffloadMetadata {
+    pub fn new(payload_size: u64, mode: TransferKind) -> Self {
+        OffloadMetadata { payload_size, mode }
+    }
+
+    pub fn from_ty<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>) -> Self {
+        OffloadMetadata { payload_size: get_payload_size(tcx, ty), mode: TransferKind::Both }
```

Comment:
If you already have the code here, I would add a small check for `&` or byVal (implies Mode ToGPU), vs `&mut` (implies Both).

In the future we would hope to analyze the & or byval case more, if we never read from it (before writing) then we could use a new mode 4, which allocates directly on the gpu.